dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1183 @@
+				},
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		reqs, err := workerToRequests(tc.w)
+		require.NoError(t, err, fmt.Sprintf("case %d failed", i))
+		require.Equal(t, len(tc.expected), len(reqs), fmt.Sprintf("case %d failed", i))
+		require.Equal(t, tc.expected, reqs, fmt.Sprintf("case %d failed", i))
+	}
+}
+
+func TestChainSync_validateResponse(t *testing.T) {
+	t.Parallel()
+
+	cs := newTestChainSync(t)
+	ctrl := gomock.NewController(t)
+	mockBlockState := NewMockBlockState(ctrl)
+	mockBlockState.EXPECT().HasHeader(common.Hash{}).Return(true, nil).Times(4)
+	cs.blockState = mockBlockState
+
+	err := cs.validateResponse(nil, nil, "")
+	require.Equal(t, errEmptyBlockData, err)
+
+	req := &network.BlockRequestMessage{
+		RequestedData: network.RequestedDataHeader,
+	}
+
+	resp := &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Header: &types.Header{
+					Number: 1,
+				},
+				Body: &types.Body{},
+			},
+			{
+				Header: &types.Header{
+					Number: 2,
+				},
+				Body: &types.Body{},
+			},
+		},
+	}
+
+	hash := (&types.Header{
+		Number: 2,
+	}).Hash()
+	err = cs.validateResponse(req, resp, "")
+	require.Equal(t, errResponseIsNotChain, err)
+	require.True(t, cs.pendingBlocks.hasBlock(hash))
+	cs.pendingBlocks.removeBlock(hash)
+
+	parent := (&types.Header{
+		Number: 1,
+	}).Hash()
+	header3 := &types.Header{
+		ParentHash: parent,
+		Number:     3,
+	}
+	resp = &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Header: &types.Header{
+					Number: 1,
+				},
+				Body: &types.Body{},
+			},
+			{
+				Hash:          header3.Hash(),
+				Header:        header3,
+				Body:          &types.Body{},
+				Justification: &[]byte{0},
+			},
+		},
+	}
+
+	hash = (&types.Header{
+		ParentHash: parent,
+		Number:     3,
+	}).Hash()
+	err = cs.validateResponse(req, resp, "")
+	require.Equal(t, errResponseIsNotChain, err)
+	require.True(t, cs.pendingBlocks.hasBlock(hash))
+	bd := cs.pendingBlocks.getBlock(hash)
+	require.NotNil(t, bd.justification)
+	cs.pendingBlocks.removeBlock(hash)
+
+	parent = (&types.Header{
+		Number: 2,
+	}).Hash()
+	resp = &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Header: &types.Header{
+					Number: 2,
+				},
+				Body: &types.Body{},
+			},
+			{
+				Header: &types.Header{
+					ParentHash: parent,
+					Number:     3,
+				},
+				Body: &types.Body{},
+			},
+		},
+	}
+
+	err = cs.validateResponse(req, resp, "")
+	require.NoError(t, err)
+	require.False(t, cs.pendingBlocks.hasBlock(hash))
+
+	req = &network.BlockRequestMessage{
+		RequestedData: network.RequestedDataJustification,
+	}
+	resp = &network.BlockResponseMessage{
+		BlockData: []*types.BlockData{
+			{
+				Justification: &[]byte{0},
+			},
+		},
+	}
+
+	err = cs.validateResponse(req, resp, "")
+	require.NoError(t, err)
+	require.False(t, cs.pendingBlocks.hasBlock(hash))
+}
+
+func TestChainSync_validateResponse_firstBlock(t *testing.T) {
```

Comment:
Could you merge this and `TestChainSync_validateResponse` together with subtests?

Reply:
Refactored TestChainSync_validateResponse to include this.